### PR TITLE
DAOS-4838 test: Verify container user security with ACL

### PR DESCRIPTION
**Pull request by dinghwah**
Verify container user security:
            w  - set_container_attribute or data
            r  - get_container_attribute or data
            T  - set_container_property
            t  - get_container_property
            a  - get_container_acl_list
            A  - update_container_acl
            o  - set_container_owner
            d  - destroy_container

### CODE REVIEW

File: src/tests/ftest/util/pool_security_test_base.py [hunk 1]
```diff
@@ -162,6 +163,189 @@ def verify_daos_pool_result(self, result, action, expect, err_code):
...
+    def setup_container_acl_and_permission(
+        self, user_type, user_name, perm_type, perm_action):
+        """Setup container acl and permissions.
+
+        Args:
+            user_type (str): Conatiner user_type.
+            user_name (str): Conatiner user_name.
+            perm_name (str): Conatiner permission type:
+                             (attribute, property, acl, or ownership)
+            user_name (str): Conatiner permission read write action.
```
Comment (shimizukko) [inline src/tests/ftest/util/pool_security_test_base.py:320]: There are two user_name. There's no perm_type and perm_action.

**Review by shimizukko — DISMISSED** [review]
Looks good to me

File: src/tests/ftest/util/pool_security_test_base.py (hunk 1 above)
Comment (dinghwah) [reply src/tests/ftest/util/pool_security_test_base.py]: Thanks for the comment. Please see the new commit.

**Comment by dinghwah** [comment]
15 failures on the functional test were not related to the commit, daos_vol mpi test failures also observed on the master run #2805 (09/03/2020).

File: src/tests/ftest/util/pool_security_test_base.py [hunk 2]
```diff
@@ -317,7 +317,7 @@ def setup_container_acl_and_permission(
             user_name (str): Conatiner user_name.
             perm_name (str): Conatiner permission type:
```
Comment (shimizukko) [inline src/tests/ftest/util/pool_security_test_base.py:318]:
```suggestion
            perm_type (str): Conatiner permission type:
```

**Review by shimizukko — DISMISSED** [review]
LGTM

File: src/tests/ftest/security/container_security_acl.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            "(4.4)Verify container_attribute: read, expect: %s", expect_read)
+        self.verify_cont_rw_property("read", expect_read)
+
+        #(5)Verify container permissions aA, rw-acl
+        permission_type = "acl"
+        self.log.info("(5)Verify container permission aA, rw-acl ")
+        self.log.info(
+            "(5.1)Update container-acl %s, %s, permission_type: %s with %s",
+            user_type, self.current_user, permission_type, cont_permission)
+        expect = "pass"  #User who created the container has full acl access.
```
Comment (kjacque) [inline src/tests/ftest/security/container_security_acl.py:137]: You could change the owner to a different user in order to verify the case where the user is NOT the owner, when get/set-ACL access is denied.

File: src/tests/ftest/security/container_security_acl.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            "write", expect, secTestBase.acl_entry(
+                user_type, self.current_user, cont_permission))
+        self.log.info("(5.3)Verify container_acl: read, expect: %s", expect)
+        self.verify_cont_rw_acl("read", expect)
+
+        #(6)Verify container permission o, set-owner
+        self.log.info("(6)==>Verify container permission o, set-owner")
+        permission_type = "ownership"
+        expect = "deny"
+        if "w" in cont_permission:
```
Comment (kjacque) [inline src/tests/ftest/security/container_security_acl.py:152]: should be "o"?

File: src/tests/ftest/security/container_security_acl.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        self.verify_cont_set_owner(expect, test_user+"@", test_group+"@")
+        self.log.info("(6.3)Verify container_ownership: read, expect: %s",
+            expect)
+        self.verify_cont_rw_property("read", expect)
+
+        #(7)Verify container permission d, delete
+        self.log.info(" ")
+        self.log.info("(7)==>Verify container d, delete ")
+        expect = "pass"
+        self.verify_cont_delete(expect)
```
Comment (kjacque) [inline src/tests/ftest/security/container_security_acl.py:171]: Delete is an interesting case. I think you are creating the pool with default permissions above (tcd or rw). If a user has "d" permission on the pool, they can delete any container in it. If you want to test the container delete permission, I would suggest creating the pool with "tc" permissions for the owner and group. Then they can only delete the container if they have "d" permission on the container.

So possible container delete cases are:
- Pool permission includes "d", container permission includes "d": ALLOWED
- Pool permission includes "d", container permission does NOT include "d": ALLOWED
- Pool permission does NOT include "d", container permission includes "d": ALLOWED
- Pool permission does NOT include "d", container permission does NOT include "d": DENIED

File: src/tests/ftest/util/pool_security_test_base.py (hunk 2 above)
Comment (dinghwah) [reply src/tests/ftest/util/pool_security_test_base.py]: Updated. Thanks.

File: src/tests/ftest/security/container_security_acl.py (hunk 3 above)
Comment (dinghwah) [reply src/tests/ftest/security/container_security_acl.py:138]: Thanks for the comment.
I have added a new test at step (6-4), after set-owner and verify the "deny" of acl-set.

File: src/tests/ftest/security/container_security_acl.py (hunk 4 above)
Comment (dinghwah) [reply src/tests/ftest/security/container_security_acl.py:152]: "w" is read from test.yaml file (var cont_permission) which mux with 4 options: none, "r", "w", and "rw" for each of the rw, tT, aA and o (w only) tests.
On set-owner "o" test, test script perform the correct acl-set with "o".

File: src/tests/ftest/security/container_security_acl.py (hunk 5 above)
Comment (dinghwah) [reply src/tests/ftest/security/container_security_acl.py]: Thanks for the comment.
I have added 4 tests for container delete at step(7),

**Review by shimizukko — APPROVED** [review]
LGMT

**Comment by dinghwah** [comment]
(1) 3 container acl tests failed on Leap15 are due to DAOS-5635.
(2) The other 4 failures on degrade-mode and io-test are due to DAOS-5452 and DAOS-5488, not related to this PR commit.

**Review by kjacque — APPROVED** [review]
LGTM

**Comment by sylviachanoiyee** [comment]
@dinghwah , the tests failed. Is this expected?
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3367/11/testReport/(root)/DaosContainterSecurityTest/Test___Functional_on_Leap_15___1___security_container_security_acl_py_DaosContainterSecurityTest_test_container_user_acl_container_no_permission_hosts_pool_server_config_0a78/

**Review by sylviachanoiyee — COMMENTED** [review]
The tests failed in CI. Is this expected?
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3367/11/testReport/(root)/DaosContainterSecurityTest/Test___Functional_on_Leap_15___1___security_container_security_acl_py_DaosContainterSecurityTest_test_container_user_acl_container_no_permission_hosts_pool_server_config_0a78/

**Comment by dinghwah** [comment]
Per comment earlier,  3 container acl tests failed on Leap15 are due to DAOS-5635.
All tests on CentOS7 are passing.

**Comment by mjmac** [comment]
Hmm. I don't think we can land this without a resolution for DAOS-5635. @dinghwah, can you add any more information to that ticket to help us understand what the problem might be and who could help fix it? I'm wondering why only this test is failing on Leap15, for example.

**Comment by dinghwah** [comment]
The tests for container-acl user, includes positive and negative tests.
For negative tests, test script is looking for "-1001" return after daos container-acl command.
From the test-log of Leap15, the error message was redirected to stdout and stderr empty.
The tests did not perform the negative test and looking for the return of stderr would not see this issue.

**Review by kjacque — DISMISSED** [review]

**Review by shimizukko — APPROVED** [review]
LGTM

**Review by kjacque — DISMISSED** [review]

**Comment by dinghwah** [comment]
Script has been updated to parse the output of stdout and stderr.
All container_user_acl tests passed.
Errors in pr-head and unit-test are not related to this commit.

**Review by shimizukko — APPROVED** [review]
LGTM

**Review by kjacque — APPROVED** [review]

**Comment by dinghwah** [comment]
(1)All container_user_acl tests passed, on both Leap15 and CentOS7.
(2)6 functional failed tests are due to:
    DAOS-5500: pool evict test timeout
    DAOS-5416: Pool Version Error: After exclude on osa_online_reintegration test
    DAOS-5559: creating the /tmp/daos_dfuse/ dfuse mount point
    DAOS-5452: daos_core_test-rebuild 0-10: daos_io_server: corrupted size

**Pull request merged**